src/Components/Components/src/ComponentFactory.cs

```diff
@@ -95,16 +96,17 @@ private static void PerformPropertyInjection(IServiceProvider serviceProvider, I
     private static Action<IServiceProvider, IComponent> CreatePropertyInjector([DynamicallyAccessedMembers(Component)] Type type)
     {
         // Do all the reflection up front
-        List<(string name, Type propertyType, PropertySetter setter)>? injectables = null;
+        List<(string name, Type propertyType, PropertySetter setter, object? serviceKey)>? injectables = null;
         foreach (var property in MemberAssignment.GetPropertiesIncludingInherited(type, _injectablePropertyBindingFlags))
         {
-            if (!property.IsDefined(typeof(InjectAttribute)))
+            var injectAttribute = property.GetCustomAttribute<InjectAttribute>();
+            if (injectAttribute is null)
             {
                 continue;
             }
 
             injectables ??= new();
-            injectables.Add((property.Name, property.PropertyType, new PropertySetter(type, property)));
+            injectables.Add((property.Name, property.PropertyType, new PropertySetter(type, property), injectAttribute.Key));
```

Comment:
Looking at the caller, it seems you could pass in `IServiceProvider` and do the `IServiceProvider is IServiceProviderIsKeyedService keyedServiceProvider && keyedServiceProvider.IsKeyedService(...)` check and throw early if keyed services aren't supported.

Reply:
Good point. The thing is, though, that `ComponentFactory` in its current state isn't limited to using a single `IServiceProvider` instance, so it might not always be correct to verify the existence of a service on the first `IServiceProvider` and reuse that result for future `IServiceProvider`s.

Also, it might happen that we create/cache a property injector but never invoke it, in which case we throw an exception unnecessarily. For example, it's possible that a component could have an interactive render mode that disables prerendering, so the component never actually gets rendered by the endpoint renderer (and we hit that final `else` block in `InstantiateComponent()`). In this case, the component might require a service unavailable during prerendering but available during interactive rendering, and we would throw an exception unnecessarily.

So, I'm thinking it might be best to leave the validation at the point when `[Inject]` properties are populated. Does that seem reasonable?